// mismatched/dist/src/examples/eg.micro.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const assertThat_1 = require("../assertThat");
const match_1 = require("../match");
describe("Object-matching Examples", () => {
    const actual = { name: "hamcrest", address: { number: 3, street: "Oak St", other: [1, 2] } };
    describe("assertThat():", () => {
        it('Full object match', () => {
            (0, assertThat_1.assertThat)(actual)
                .is({ name: "hamcrest", address: { number: 3, street: "Oak St", other: [1, 2] } });
        });
        describe("Partial object match", () => {
            it('Do not care about one field', () => {
                (0, assertThat_1.assertThat)(actual)
                    .is({
                    name: "hamcrest",
                    address: { number: 3, street: "Oak St", other: match_1.match.any() }
                });
            });
            it('Matching optionally on several fields', () => {
                (0, assertThat_1.assertThat)(actual)
                    .is({
                    name: match_1.match.anyOf(["hamcrest", "tsDiffMatcher"]),
                    address: {
                        number: match_1.match.number.greater(0),
                        street: "Oak St",
                        other: match_1.match.ofType.array()
                    }
                });
            });
        });
    });
    describe("validateThat():", () => {
        it('Full object validation', () => {
            (0, assertThat_1.assertThat)(actual)
                .is({
                name: match_1.match.ofType.string(),
                address: {
                    number: match_1.match.number.greater(0),
                    street: match_1.match.ofType.string(),
                    other: match_1.match.array.every(match_1.match.ofType.number())
                }
            });
        });
    });
});
