// mismatched/dist/src/prettyPrint/tile/ObjectTile.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FieldTile = exports.ObjectTile = void 0;
const printSequenceOfTiles_1 = require("./printSequenceOfTiles");
class ObjectTile {
    constructor(elements) {
        this.elements = elements;
        this.stringLength = Math.max(2, elements.reduceRight((accumulator, current) => accumulator + current.stringLength + 2, 0));
        this.complexity = elements.reduceRight((accumulator, current) => accumulator + current.complexity, 2);
    }
    render(appender) {
        (0, printSequenceOfTiles_1.printSequenceOfTiles)(this.elements, "{", "}", this.stringLength, appender, this.complexity);
    }
}
exports.ObjectTile = ObjectTile;
class FieldTile {
    constructor(key, value) {
        this.key = key;
        this.value = value;
        this.stringLength = key.length + 2 + value.stringLength;
        this.complexity = 1 + value.complexity;
    }
    render(appender) {
        appender.adds([this.key, ": "]);
        this.value.render(appender);
    }
}
exports.FieldTile = FieldTile;
