// mismatched/dist/src/prettyPrint/tile/PseudoCallTile.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PseudoCallTile = void 0;
const printSequenceOfTiles_1 = require("./printSequenceOfTiles");
class PseudoCallTile {
    constructor(callName, args, withSquareBrackets = false) {
        this.callName = callName;
        this.args = args;
        this.withSquareBrackets = withSquareBrackets;
        let argsStringLength = args ? args.reduceRight((accumulator, current) => accumulator + current.stringLength + 2, callName.length) : 0;
        this.stringLength = Math.max(3, argsStringLength);
        this.complexity = args ? args.reduceRight((accumulator, current) => accumulator + current.complexity, 3) : 1;
    }
    render(appender) {
        const newLine = this.stringLength > appender.remaininglineWidth || appender.tooComplex(this.complexity);
        if (newLine) {
            appender.newLine();
        }
        appender.add(this.callName);
        if (this.args) {
            appender.add(this.withSquareBrackets ? "([" : "(");
            (0, printSequenceOfTiles_1.printTiles)(this.args, appender);
            appender.add(this.withSquareBrackets ? "])" : ")");
        }
    }
}
exports.PseudoCallTile = PseudoCallTile;
