// thespian/dist/src/DefinedSetUp.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SetUpType = exports.DefinedSetUp = void 0;
class DefinedSetUp {
    static details(f) {
        const fn = f.toString(); // Eg, 'f => f.foooo(2, "aaa")'
        const split = fn.split(" => "); // Eg, ['f', 'f.foooo(2, "aaa")'
        const call = split[1]; // eg, 'f.foooo(2, "aaa")'
        const openBracket = call.indexOf("(");
        const dot = call.indexOf(".");
        if (openBracket < 0) {
            return { _type: SetUpType.Property, name: call.slice(dot + 1) };
        }
        if (dot < 0 || dot > openBracket) { // Eg, (j: any) => j(match.any())
            return { _type: SetUpType.Function }; // a function
        }
        const name = call.slice(dot + 1, openBracket);
        return { _type: SetUpType.Method, name: name };
    }
}
exports.DefinedSetUp = DefinedSetUp;
var SetUpType;
(function (SetUpType) {
    SetUpType[SetUpType["Method"] = 0] = "Method";
    SetUpType[SetUpType["Property"] = 1] = "Property";
    SetUpType[SetUpType["Function"] = 2] = "Function";
})(SetUpType = exports.SetUpType || (exports.SetUpType = {}));
