// thespian/dist/src/UnsuccessfulCall.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UnsuccessfulCall = void 0;
const SuccessfulCall_1 = require("./SuccessfulCall");
class UnsuccessfulCall {
    constructor(call, matchRate, expectedTimes, actualTimes) {
        this.call = call;
        this.matchRate = matchRate;
        this.expectedTimes = expectedTimes;
        this.actualTimes = actualTimes;
    }
    describe() {
        return { call: this.call, expectedTimes: this.expectedTimes, actualTimes: this.actualTimes };
    }
    tooManyTimes() {
        return this.actualTimes > this.expectedTimes;
    }
    static make(name, matchRate, expectedArgs, expectedTimes, actualTimes) {
        return new UnsuccessfulCall((0, SuccessfulCall_1.createPseudoCall)(name, expectedArgs), matchRate, expectedTimes, actualTimes);
    }
    static makeNearMiss(name, matchResult, expectedTimes, actualTimes) {
        return new UnsuccessfulCall((0, SuccessfulCall_1.createPseudoCall)(name, matchResult.diff), matchResult.matchRate, expectedTimes, actualTimes);
    }
}
exports.UnsuccessfulCall = UnsuccessfulCall;
