// mismatched/dist/src/Colour.js
"use strict";
// See https://www.npmjs.com/package/ansi-escape-sequences
Object.defineProperty(exports, "__esModule", { value: true });
exports.Colour = void 0;
const black = "\u001b[30m";
const reset = "\u001b[0m";
class Colour {
    static black(s) {
        return black + s;
    }
    static red(s) {
        return "\u001b[31m" + s + reset;
    }
    static green(s) {
        return "\u001b[32m" + s + reset;
    }
    static yellow(s) {
        return "\u001b[33m" + s + reset;
    }
    static blue(s) {
        return "\u001b[34m" + s + reset;
    }
    static magenta(s) {
        return "\u001b[35m" + s + reset;
    }
    static cyan(s) {
        return "\u001b[36m" + s + reset;
    }
    static white(s) {
        return "\u001b[37m" + s + reset;
    }
    static gray(s) {
        return "\u001b[90m" + s + reset;
    }
    static bg_black(s) {
        return "\u001b[40m" + s + reset;
    }
    static bg_red(s) {
        return "\u001b[41m" + s + reset;
    }
    static bg_green(s) {
        return "\u001b[42m" + s + reset;
    }
    static bg_yellow(s) {
        return "\u001b[43m" + s + reset;
    }
    static bg_blue(s) {
        return "\u001b[44m" + s + reset;
    }
    static bg_magenta(s) {
        return "\u001b[45m" + s + reset;
    }
    static bg_cyan(s) {
        return "\u001b[46m" + s + reset;
    }
    static bg_white(s) {
        return "\u001b[47m" + s + reset;
    }
    static bg_gray(s) {
        return "\u001b[100m" + s + reset;
    }
}
exports.Colour = Colour;
