// mismatched/dist/src/ofType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ofType = void 0;
function isUndefined(v) {
    return typeof v === "undefined";
}
function isDefined(v) {
    return !isUndefined(v);
}
function isNull(v) {
    return v === null;
}
function isObject(v) {
    return v !== null && !isArray(v) && typeof v === "object" && !(v instanceof RegExp);
}
function isArray(v) {
    return Array.isArray(v);
}
function isFunction(v) {
    return typeof v === "function";
}
function isSet(v) {
    return v instanceof Set;
}
function isMap(v) {
    return v instanceof Map;
}
function isString(v) {
    return typeof v === 'string' || v instanceof String;
}
function isNumber(v) {
    return typeof v === 'number'; // && isFinite(value);
}
function isNaN(v) {
    return v != v; // Avoid weird JS coercing that make isNaN() and Number.isNaN() unreliable
}
function isBoolean(v) {
    return typeof v === 'boolean';
}
function isRegExp(v) {
    return v instanceof RegExp;
}
function isSymbol(v) {
    return typeof v === 'symbol';
}
function isMatcher(v) {
    return isObject(v) && !isUndefined(v.matches) && isFunction(v.matches);
}
function isError(v) {
    return v instanceof Error;
}
exports.ofType = {
    isArray,
    isSet,
    isMap,
    isBoolean,
    isError,
    isFunction,
    isMatcher,
    isNaN,
    isNumber,
    isNull,
    isObject,
    isRegExp,
    isString,
    isSymbol,
    isUndefined,
    isDefined
};
