// mismatched/dist/src/validateThat.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ValidationResult = exports.validateThat = void 0;
const DiffMatcher_1 = require("./matcher/DiffMatcher");
const matchMaker_1 = require("./matchMaker/matchMaker");
const assertThat_1 = require("./assertThat");
const PrettyPrinter_1 = require("./prettyPrint/PrettyPrinter");
const ofType_1 = require("./ofType");
function validateThat(actual) {
    return new Validator(actual);
}
exports.validateThat = validateThat;
class Validator {
    constructor(actual) {
        this.actual = actual;
    }
    satisfies(expected) {
        (0, assertThat_1.ensureNotFunction)(this.actual);
        const matcher = (0, matchMaker_1.matchMaker)(expected);
        const mismatched = [];
        matcher.mismatches(new DiffMatcher_1.ContextOfValidationError(), mismatched, this.actual);
        return new ValidationResult(mismatched.map(m => (ofType_1.ofType.isString(m) ? m :
            PrettyPrinter_1.PrettyPrinter.make(500, 5000).render(m))));
    }
    is(expected) {
        return this.satisfies(expected);
    }
}
class ValidationResult {
    constructor(errors) {
        this.errors = errors;
        this.passed = () => this.errors.length === 0;
    }
}
exports.ValidationResult = ValidationResult;
