// mismatched/dist/src/matcher/DescribeMatcher.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DescribeMatcher = void 0;
const DiffMatcher_1 = require("./DiffMatcher");
const MatchResult_1 = require("../MatchResult");
const __1 = require("..");
class DescribeMatcher extends DiffMatcher_1.DiffMatcher {
    constructor(matcher, description) {
        super();
        this.matcher = matcher;
        this.description = description;
    }
    static make(matcher, description) {
        return new DescribeMatcher((0, __1.matchMaker)(matcher), description);
    }
    mismatches(context, mismatched, actual) {
        const matchResult = this.matcher.mismatches(context, [], actual);
        if (matchResult.passed()) {
            return MatchResult_1.MatchResult.good(matchResult.compares);
        }
        mismatched.push(this.description(actual, context.context));
        return matchResult; // Has no impact on assertThat() results
    }
    describe() {
        return "describe";
    }
}
exports.DescribeMatcher = DescribeMatcher;
