// mismatched/dist/src/matcher/DiffFieldMatcher.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DiffFieldMatcher = void 0;
const DiffMatcher_1 = require("./DiffMatcher");
const matchMaker_1 = require("../matchMaker/matchMaker");
const ObjectKeyMatcher_1 = require("./ObjectKeyMatcher");
const BindMatcher_1 = require("./BindMatcher");
const allKeys_1 = require("../allKeys");
class DiffFieldMatcher extends DiffMatcher_1.DiffMatcher {
    constructor(fieldName, matcher) {
        super();
        this.fieldName = fieldName;
        this.matcher = matcher;
        this.specificity = matcher.specificity;
    }
    mismatches(context, errors, actual) {
        return this.matcher.mismatches(context.add("." + this.fieldName.toString()), errors, actual[this.fieldName]);
    }
    describe() {
        return { [this.fieldName]: this.matcher.describe() };
    }
    isKey() {
        return this.matcher instanceof ObjectKeyMatcher_1.ObjectKeyMatcher;
    }
    isBind() {
        return this.matcher instanceof BindMatcher_1.BindMatcher;
    }
    static make(fieldName, expected) {
        return new DiffFieldMatcher(fieldName, (0, matchMaker_1.matchMaker)(expected));
    }
    static makeAll(obj) {
        return (0, allKeys_1.allKeys)(obj).map(key => new DiffFieldMatcher(key, (0, matchMaker_1.matchMaker)(obj[key])));
    }
}
exports.DiffFieldMatcher = DiffFieldMatcher;
