// mismatched/dist/src/matcher/ErrorMatcher.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ErrorMatcher = void 0;
const DiffMatcher_1 = require("./DiffMatcher");
const MatchResult_1 = require("../MatchResult");
const matchMaker_1 = require("../matchMaker/matchMaker");
const instanceOfMatcher_1 = require("./instanceOfMatcher");
class ErrorMatcher extends DiffMatcher_1.DiffMatcher {
    constructor(matcher) {
        super();
        this.matcher = matcher;
        this.specificity = matcher.specificity;
    }
    mismatches(context, mismatched, actual) {
        const typeMatchResult = instanceOfMatcher_1.instanceOfMatcher.instanceOf(Error).mismatches(context, mismatched, actual);
        if (!typeMatchResult.passed()) {
            return typeMatchResult;
        }
        const matchResult = this.matcher.mismatches(context, mismatched, actual.message);
        if (matchResult.passed()) {
            return MatchResult_1.MatchResult.good(1);
        }
        return MatchResult_1.MatchResult.wasExpected(actual, this.describe(), 1, 0);
    }
    describe() {
        return { errorMessage: this.matcher.describe() };
    }
    static make(matcher) {
        return new ErrorMatcher((0, matchMaker_1.matchMaker)(matcher));
    }
}
exports.ErrorMatcher = ErrorMatcher;
