// mismatched/dist/src/matcher/MappedMatcher.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MappedMatcher = void 0;
const DiffMatcher_1 = require("./DiffMatcher");
const MatchResult_1 = require("../MatchResult");
const matchMaker_1 = require("../matchMaker/matchMaker");
class MappedMatcher extends DiffMatcher_1.DiffMatcher {
    constructor(map, matcher, description) {
        super();
        this.map = map;
        this.matcher = matcher;
        this.description = description;
    }
    static make(map, matcher, description) {
        return new MappedMatcher(map, (0, matchMaker_1.matchMaker)(matcher), description);
    }
    mismatches(context, mismatched, actual) {
        const matchResult = this.matcher.mismatches(context, mismatched, this.map(actual));
        if (matchResult.passed()) {
            return MatchResult_1.MatchResult.good(1);
        }
        return MatchResult_1.MatchResult.wasExpected(actual, this.describe(), 1, 0);
    }
    describe() {
        return { mapped: { description: this.description, matcher: this.matcher.describe() } };
    }
}
exports.MappedMatcher = MappedMatcher;
