// mismatched/dist/src/matcher/NotMatcher.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NotMatcher = void 0;
const DiffMatcher_1 = require("./DiffMatcher");
const MatchResult_1 = require("../MatchResult");
const matchMaker_1 = require("../matchMaker/matchMaker");
const Mismatched_1 = require("./Mismatched");
class NotMatcher extends DiffMatcher_1.DiffMatcher {
    constructor(matcher) {
        super();
        this.matcher = matcher;
        this.specificity = matcher.specificity;
    }
    mismatches(context, mismatched, actual) {
        const matchResult = this.matcher.matches(actual);
        if (matchResult.passed()) {
            mismatched.push(Mismatched_1.Mismatched.makeExpectedMessage(context, actual, this.describe()));
            return MatchResult_1.MatchResult.wasExpected(actual, this.describe(), 1, 0);
        }
        return MatchResult_1.MatchResult.good(1);
    }
    describe() {
        return { not: this.matcher.describe() };
    }
    static make(matcher) {
        return new NotMatcher((0, matchMaker_1.matchMaker)(matcher));
    }
}
exports.NotMatcher = NotMatcher;
